handbook/python-dataplugins/python-dataplugins.md

```diff
@@ -0,0 +1,308 @@
+
+   1. assign an empty array to the channel values property:
+
+      ```python
+      ...
+      'groups': [{
+         ...
+         'channels': [{
+            ...
+            'values': [],
+         }]
+      }]
+      ```
+
+   1. outsource the functionality to load the bulk data values in a separate function.
+   The function has the following definition:
+
+      ```python
+      def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+         """
+            Returns a value array of the correct data type specified in the tdm dictionary
+         """
+      ```
+
+<!-- markdownlint-disable -->
+<details>
+<summary>Example Code</summary>
+<!-- markdownlint-enable -->
+
+```python
+def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+   dataType = self.tdm_tree["groups"][grp_index]["channels"][chn_index]["type"]
+   values = []
+   for row in self.data:
+      value = row[self.channelNames[chn_index]]
+      values.append(value)
+   return values[numberToSkip:numberToTake+numberToSkip]
+```
+
+</details>
+
+The client applications are calling that function to retrieve channel values (or
+a subset of values). The DataPlugin needs to implement the function to return the
+values for a given group and channel index. It also needs to ensure only the
+correct subset of values is returned for a given `numberToSkip` and `numberToTake`.
+
+Additionally, a callback function to retrieve the channel length must be
+implemented. In the simple case where all channels have the same length, it can
+simply return a constant.
+
+```python
+def read_channel_length(self, grp_index, chn_index):
+   """
+      Returns the channel length as an Integer for a given group and channel index
+   """
+   return 10
+```
+
+See full example: [csv-read-with-callback-loading](https://github.com/ni/systemlink-operations-handbook/tree/master/examples/python-dataplugin-examples/csv-read-with-callback-loading)
+
+## Error Handling
+
+Python DataPlugins can raise errors in all callback functions. The raised error
+will be transported to DIAdem, LabVIEW or SystemLink DataFinder.
+
+```python
+def read_store(self, parameter):
+   ...
+   raise Exception("Leave read_store with exception")
+```
+
+### Not My File
+
+A special case is "Not My File". This error should be raised when the DataPlugin
+detects that the file to be opened, is not suited for this DataPlugin. This
+special error can be raised only in the `read_store` function by returning `None`:
+
+```python
+def read_store(self, parameter):
+   ...
+   if notMyFile:
+      return None
+```
+
+## Export DataPlugin
+
+Export Python DataPlugins to make them available on other systems. Use DIAdem to
+export a DataPlugin as a URI file.
+
+<figure>
+   <img src="../../img/pydp-diadem_export.png" width="600" />
+   <figcaption>Exporting DataPlugins in DIAdem</figcaption>
+</figure>
+
+## Known Limitations
+
+### Numpy and Pandas

```

Comment:
NumPy I think is the correct notation (also in other places in this documentation)